### DISCUSSION TITLE
A Tree-Sitter parser for Slint

### DISCUSSION THREAD

**Discussion by jrmoulton** [slint-ui/slint#970] [Show and tell]
Hey everyone!

I built a tree sitter parser for Slint! Its intended use is for neovim and currently supports syntax highlighting and indentation. Link [here](https://github.com/jrmoulton/tree-sitter-slint)

### Without Tree Sitter
<img width="649" alt="no_tree_sitter" src="https://user-images.githubusercontent.com/13090441/154774858-9d3f7e9a-8e3b-4164-9218-5b8418960d52.png">

### With Tree Sitter
<img width="642" alt="with_tree_sitter" src="https://user-images.githubusercontent.com/13090441/154774873-4e0fc34a-d55b-4808-a57e-3cb620e4371f.png">

This is the first time I've ever written a tree-sitter parser so the implementation could still use some work but highlighting works great!

**Comment by jrmoulton** [comment]
This image wouldn't compile but it shows working highlighting for a lot (if not all of) the language features.

<img width="566" alt="all-features" src="https://user-images.githubusercontent.com/13090441/154791457-6f059a04-8150-4578-a9cc-7ce13d49cc73.png">

**Reply by hunger** [reply]
This is seriously cool!

Is it in a state where you feel comfortable to have create a PR for `nvim-treesitter` yet?

The slint language server is already known to the `nvim-lspconfig` plugin and a PR to `nvim-lsp-instsller` is open. It would be cool to have all the LSP and tree-sitter bits work out of the box for neovim!

You probably still need the basic language support plugin... which I should probably just open a PR for the `vim-polyglott` plugin for. Then we should have slint support in lots of nvim configurations without the need to even install any extra plugin.

**Reply by jrmoulton** [reply]
Thanks!
I'm not quite to a point where I want to create a pull request for nvim-treesitter just yet. I haven't added any unit tests but I'll be working on that and as soon as I feel like I have a decent test coverage I'll create a pull request.

**Comment by hunger** [comment]
What is the config you use to register your Slint parser with `nvim-treesitter`? I would really like to give this a spin:-)

**Reply by jrmoulton** [reply]
Specifically for slint I have 
```lua 
local parser_config = require "nvim-treesitter.parsers".get_parser_configs()
parser_config.slint = {
    install_info = {
        url = "https://github.com/jrmoulton/tree-sitter-slint",
        files = {"src/parser.c"}
    },
    filetype = "slint",
}

```
And until I get a PR merged into nvim-treesitter it is also necessary to copy the files from the queries folder from tree-sitter-slint to a local config/nvim/queries/slint folder that you have to create

**Comment by hunger** [comment]
Would it maybe be interesting for you to move this work into the slint-ui organization? I just moved the vim-slint plugin yesterday to make that a bit more official.

You would of course retain copyright and write access and everything and can maintain it as before, but this would require the CLA to be in effect for the tree-sitter code as well.

**Reply by jrmoulton** [reply]
Yeah I think that would be a great path forward for the project! I think that would be the best way for it to be adopted and get the most support. And I would be perfectly ok with major revisions being done.

**Reply by ogoffart** [reply]
Ideally, we should have it in the main slint repository. Right now there is some editor stuff on tools/lsp/* and the vscode_extension in its own folder.
We should move the editor specific stuff to a new folder in editors/ or something like that
